### PR TITLE
#963 capture bazel build info into files, print info when return code is not zero

### PR DESCRIPTION
**Pull request by caofangkun**
image

### CODE REVIEW

**Comment by kramasamy** [comment]
@maosongfu and @nlu90 - can you please review this?

**Comment by objmagic** [comment]
Is Apache 2.0 the correct license to use? Also note that other scripts under ``script/travis`` do not have any license in header.

File: scripts/travis/save-logs.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import sys
+import subprocess
+import mmap
+import os
```
Comment (nlu90) [inline scripts/travis/save-logs.py]: please sort imports

File: scripts/travis/save-logs.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                if fm[i] == '\n':
+                    n -= 1
+                    if n == -1:
+                        break
+            return fm[i + 1 if i else 0:].splitlines()
+        finally:
+            fm.close()
+
+def main(file, cmd):
+    print cmd, "writing to", file
```
Comment (nlu90) [inline scripts/travis/save-logs.py]: `print "%s writing to: %s" % (cmd, file)` is better

File: scripts/travis/save-logs.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+#  limitations under the License.
+
+import sys
+import subprocess
+import mmap
+import os
+
+from datetime import datetime, timedelta
+
+def tail(filename, n):
```
Comment (nlu90) [inline scripts/travis/save-logs.py]: Can we call the "tail" shell cmd in python to do the task?

**Comment by objmagic** [comment]
this may come as a surprise, but we require Python indentation level to be **2**. It would be good if you are willing to adopt our coding style.

File: scripts/travis/save-logs.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                    n -= 1
+                    if n == -1:
+                        break
+            return fm[i + 1 if i else 0:].splitlines()
+        finally:
+            fm.close()
+
+def main(file, cmd):
+    print cmd, "writing to", file
+    out = open(file, "w")
```
Comment (nlu90) [inline scripts/travis/save-logs.py]: `with open(file, "w") as out:`

File: scripts/travis/save-logs.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    count = 0
+    process = subprocess.Popen(cmd,
+                           stderr=subprocess.STDOUT,
+                           stdout=subprocess.PIPE)
+
+    start = datetime.now()
+    nextPrint = datetime.now() + timedelta(seconds=1)
```
Comment (nlu90) [inline scripts/travis/save-logs.py]: should we use `start` here as `nextPrint = start + timedelta(seconds=1)`?

File: scripts/travis/save-logs.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            diff = datetime.now() - start
+            sys.stdout.write("\r%d seconds %d log lines"%(diff.seconds, count))
+            sys.stdout.flush()
+            nextPrint = datetime.now() + timedelta(seconds=10)
+        out.write(line)
+        line = pout.readline()
+    out.close()
+    errcode = process.wait()
+    diff = datetime.now() - start
+    sys.stdout.write("\r%d seconds %d log lines"%(diff.seconds, count))
```
Comment (nlu90) [inline scripts/travis/save-logs.py]: this is a nitpick: diff is timedelta and diff.seconds only returns the seconds difference, it ignores higher granularity difference

File: scripts/travis/save-logs.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    print
+    print cmd, "done", errcode
```
Comment (nlu90) [inline scripts/travis/save-logs.py]: `print "\n %s finished with errcode: %d" % (cmd, errcode)`

File: scripts/travis/save-logs.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    if errcode != 0:
+       lines = tail(file, 1000)
+       for line in lines:
+           print line
```
Comment (nlu90) [inline scripts/travis/save-logs.py]: why not print the lines directly?

File: scripts/travis/save-logs.py (hunk 1 above)
Comment (caofangkun) [inline scripts/travis/save-logs.py]: fixed

File: scripts/travis/save-logs.py (hunk 2 above)
Comment (caofangkun) [inline scripts/travis/save-logs.py]: fixed

File: scripts/travis/save-logs.py (hunk 3 above)
Comment (caofangkun) [inline scripts/travis/save-logs.py]: Yes ,we can call "tail" shell. but I do not think that make any difference.

Code Sample:
```
#!/usr/bin/python
# -*- coding: utf-8 -*-

import subprocess 
filename = '/tmp/result.log'  
command='tail -1000f '+filename
popen=subprocess.Popen(command,stdout=subprocess.PIPE,stderr=subprocess.PIPE,shell=True)  
while True:  
    line=popen.stdout.readline().strip()  
	print line
```

File: scripts/travis/save-logs.py (hunk 4 above)
Comment (caofangkun) [inline scripts/travis/save-logs.py]: fixed

File: scripts/travis/save-logs.py (hunk 5 above)
Comment (caofangkun) [inline scripts/travis/save-logs.py]: fixed

File: scripts/travis/save-logs.py (hunk 6 above)
Comment (caofangkun) [inline scripts/travis/save-logs.py]: seconds difference is enough here
```
===> Starting heron build at 2016-06-22 03:11:50
['bazel', '--bazelrc=tools/travis-ci/bazel.rc', 'build', 'heron/...'] writing to heron_build.txt
1472 seconds 5946 log lines
```

File: scripts/travis/save-logs.py (hunk 7 above)
Comment (caofangkun) [inline scripts/travis/save-logs.py]: fixed

File: scripts/travis/save-logs.py (hunk 8 above)
Comment (caofangkun) [inline scripts/travis/save-logs.py]: ```lines``` is ```List```  , print it line by line will be much readable.

eg: 
* ```print lines ```  will be 
```
["INFO: Reading 'startup' options from /home/ablecao/workspace/twitter-heron/tools/travis-ci/bazel.rc: --host_jvm_args=-Xmx2500m --host_jvm_args=-Xms2500m --batch", '____(06-23 11:01:59.475) Loading package...]
```
*  ```print line by line ```  will be 

```
____(06-23 11:00:02.695) Loading package: heron/common/tests/cpp/network
____(06-23 11:00:02.735) Loading package: heron/cli/src/python
____(06-23 11:00:02.765) Loading package: heron/ui/resources
____(06-23 11:00:04.142) Loading...
____(06-23 11:00:04.332) Loading package: tools/defaults
____(06-23 11:00:04.396) Loading package: @bazel_tools//tools/objc
____(06-23 11:00:04.449) Loading package: third_party/java

```

Comment (nlu90) [inline scripts/travis/save-logs.py]: `print '\n'.join(lines)` is simpler

File: scripts/travis/save-logs.py (hunk 3 above)
Comment (nlu90) [inline scripts/travis/save-logs.py]: @caofangkun If the `tail` cmd can function well here, please use `tail`. It's easier to understand and maintain in the future.

For the example code you write in the reply, do we really need the `while` loop?

**Comment by kramasamy** [comment]
this looks good to me. @maosongfu and @nlu90 - are you fine with merging it?

File: scripts/travis/save-logs.py (hunk 8 above)
Comment (kramasamy) [inline scripts/travis/save-logs.py]: @caofangkun - any thoughts about this feedback?

Comment (caofangkun) [inline scripts/travis/save-logs.py]: @kramasamy ```print '\n'.join(lines)``` is better, I have fixed this as request. :)

**Comment by nlu90** [comment]
once the `"\n".join(lines)` change is made, it will be good to be merged

**Comment by caofangkun** [comment]
Hi @kramasamy  @nlu90 
I'm sorry , I deleted the fork yesterday， now the pull request came from an "unknown repository".
So I re-create the pull-request as  #982 , and I will delete this PR.

**Pull request closed without merging**